### PR TITLE
CoordinateReferences

### PR DESCRIPTION
**Pull request by j042**
# Overview
- you can now have direct access to underlying cpp coordinates. Don't really have to use it all the time, but if you have a certain operation that uses high dim / high order splines, and need to change only control points, this may save you some time.  Doesn't support inplace operations. 
- Only create local trackable copies on demand. sometimes, there are times where we never need them
- add small set of test utils in tests/common.py

## Addressed issues
*  there are some use cases where cps change a lot, but this always recreates core spline and for complex splines, that's expensive (especially with the approach of SplineLib's creation).
* copied properties would always have modified flags

## Checklists
* [x] Documentations are up-to-date.
* [ ] Added example(s)
* [x] Added test(s)

### CODE REVIEW

**Review by jzwar — APPROVED** [review]
LGTM, thanks for putting in the work. It took me a while to figure out why you put in all the property changes... We need to be careful to not make the code to complicated so it stays maintainable.

File: splinepy/spline.py [hunk 1]
```diff
@@ -155,6 +155,137 @@ def intersection(cls, *splines):
...
+    Helper class to core.CoordinateReferences. Allows direct
+    access to cpp's control points.
+    Named `coordinate`, since cpp core splines store weighted control points
+    for rational splines.
+    Extends functionality to provide numpy's convenient __getitem__.
+    """
+
+    __slots__ = ("core", "spline", "id_lookup", "apply_weight")
+
+    def __init__(self, core_obj, spline, apply_weight=False):
```
[the remark this answers is not in the document]
Comment (j042) [reply splinepy/spline.py:169]: it is just an attribute, so it should behave like a `property` and `setter`. is that what you mean?

File: cpp/splinepy/splines/bspline.hpp [hunk 2]
```diff
@@ -220,6 +220,25 @@ class BSpline : public splinepy::splines::SplinepyBase,
...
+    auto ref_coordinates =
+        std::make_shared<SplinepyBase::CoordinateReferences_>();
+    auto& ref_coords = *ref_coordinates;
+
+    // get ref
+    auto& coordinates = Base_::vector_space_->GetCoordinates();
+    ref_coords.reserve(coordinates.size());
+    for (auto& control_point : coordinates) {
+      for (auto& cp : control_point) {
+        ref_coords.emplace_back(RefHolder{static_cast<double&>(cp)});
```
[the remark this answers is not in the document]
Comment (j042) [reply cpp/splinepy/splines/bspline.hpp:236]: You mean this is more expensive than recreating the spline?

File: splinepy/spline.py [hunk 3]
```diff
@@ -155,6 +155,137 @@ def intersection(cls, *splines):
...
+        # in case scalar, and no apply_weight, use direct version
+        call_setitem = True
+        if isinstance(value, (int, float)):
+            if not self.apply_weight:
+                self.core.broadcast_scalar(ids, value)
+                call_setitem = False
+            else:
+                value = [value] * ids.size
+
+        # currently we will only support matching size assignment
```
[the remark this answers is not in the document]
Comment (j042) [reply splinepy/spline.py:220]: setting a single coordinate to multiple coordinate is not supported.
```
spl.coordinate_references[[1,2,3]] = [4., 5., 6.] # can't 
```

**Pull request merged**